### ISSUE TITLE
default we should prune blocks instead of archive blocks

### ISSUE THREAD

**Issue by garyyu** [GrinPlusPlus/GrinPlusPlus#63]
It looks like it never prune blocks from local database,  but we only need keep one week blocks:
```
pub const CUT_THROUGH_HORIZON: u32 = WEEK_HEIGHT as u32;
```

There's no need to keep older blocks if no explicit configuration to use it as "archive" node.

```
$ curl -0 -XGET http://127.0.0.1:13413/v1/status
{
   "chain" : {
      "hash" : "16c2986db38dbb48dd242046951af6ce5bb67e87c4593677b1f8db21b2bd8d26",
      "height" : 169698,
      "previous_hash" : "006679e47613f0ae4c2b9fef7dcd16e553f7ff438174e477d658cf1f18a560d3",
      "total_difficulty" : 16533864109
   },
   "header_height" : 169698,
   "network" : {
      "height" : 169698,
      "num_inbound" : 0,
      "num_outbound" : 15,
      "total_difficulty" : 16533864109
   },
   "protocol_version" : 1,
   "sync_status" : "FULLY_SYNCED",
   "user_agent" : "Grin++ 0.5.3"
}
```

```
$ curl -0 -XGET http://127.0.0.1:13413/v1/chain/outputs/byids?id=0889326a057edc41c2e90362de4434baeeddcc12851340ab5c3ddeb9c927985c8c
[
   {
      "commit" : "0889326a057edc41c2e90362de4434baeeddcc12851340ab5c3ddeb9c927985c8c",
      "height" : 126498,
      "mmr_index" : 371650
   }
]
```

Height `126498` is one month ago.

**Comment by garyyu** [comment]
But it seems also not a full archive mode:
```
$ curl -0 -XGET http://127.0.0.1:13413/v1/chain/outputs/byids?id=08dc5568af0333abc9d7caf321aec9cfb473179048d5f0f0138c34660d0c7ce559
null

$ curl -0 -XGET http://127.0.0.1:13413/v1/chain/outputs/byids?id=09baa6696c739b6647c843a651ef1acad67303b5b8e1999be28d82cf94cc7339f2
[
   {
      "commit" : "09baa6696c739b6647c843a651ef1acad67303b5b8e1999be28d82cf94cc7339f2",
      "height" : 9,
      "mmr_index" : 17
   }
]
```

Looks like you're saving UTXO sets both in txhashset and chain database?  for what reason?

**Comment by garyyu** [comment]
(btw, that `null` is not compatible for Jason deserialization, `[]` is better)

**Comment by DavidBurkett** [comment]
> It looks like it never prune blocks from local database, but we only need keep one week blocks:

Yea, I've had pruning implemented in a local branch for a while now, but haven't had a chance to test it thoroughly. I'll get to it eventually, but considering that Grin++ without pruning still uses less data on windows than grin with pruning, it wasn't high on my priority list.

> But it seems also not a full archive mode:

Yep, the network doesn't support full archive nodes, since there's currently no way of retrieving blocks older than the horizon.

> Looks like you're saving UTXO sets both in txhashset and chain database? for what reason?

For fast lookup by commitment. Grin does something very similar. If we implemented your idea of replacing transaction input commitments with input mmr indices, we would no longer need to duplicate that data.

> (btw, that null is not compatible for Jason deserialization, [] is better)

Yea, the JSON APIs are all a mess. I started to design new ones, but eventually gave up. I originally tried to be compatible with grin's v1 APIs, but ~75% of the way through, I found that the APIs were far from efficient, so I started writing them differently. And now everything is a mess and just needs completely redone.

**Comment by garyyu** [comment]
> For fast lookup by commitment. Grin does something very similar.

I'm sure Grin do a real prune for blocks.

For those outputs beyond the `CUT_THROUGH_HORIZON`,  they're only stored in txhashset. And indeed it's not as fast as database query, actually for txhashset, there's no way to query these outputs by commit, instead, only by leaf index:

for example:
```
curl -0 -XGET http://127.0.0.1:13413/v1/txhashset/outputs?start_index=244744\&max=1
```

So, to avoid duplicate data in database and txhashset, and considering the `fast lookup by commitment`,  perhaps we can just save `leaf_index` mapping to `commit` in database, which only need 33+8 = 41 bytes for each output, and do same prune for spent output/s.

**Comment by DavidBurkett** [comment]
> I'm sure Grin do a real prune for blocks.

My comment wasn't about pruning. It was about storing an index from commitment to mmr_index, which is exactly what Grin does: https://github.com/mimblewimble/grin/blob/master/chain/src/store.rs#L115-L121

> For those outputs beyond the CUT_THROUGH_HORIZON, they're only stored in txhashset. And indeed it's not as fast as database query, actually for txhashset, there's no way to query these outputs by commit, instead, only by leaf index:

There may not be an RPC API to query outputs by commitment, but there absolutely is a way to look them up via the database (see my above link)

> So, to avoid duplicate data in database and txhashset, and considering the fast lookup by commitment, perhaps we can just save leaf_index mapping to commit in database, which only need 33+8 = 41 bytes for each output, and do same prune for spent output/s.

Grin++ stores OutputLocation by commitment, which contains only mmr_index and block_height (very useful for performance). It's only 49 bytes per output. https://github.com/GrinPlusPlus/GrinPlusPlus/blob/64909111b7f5bb2fa424ff04b2afcb636c6f0585/include/Database/BlockDb.h#L33-L34

There's not nearly as much data duplication as you're thinking there is. Grin++ only uses 8 bytes more input than Grin, and that block height per commitment is the only additional data Grin++ stores in the database that Grin doesn't.

**Comment by garyyu** [comment]
>  Grin++ only uses 8 bytes more input than Grin, and that block height per commitment is the only additional data Grin++ stores in the database that Grin doesn't.

Wow!  `block height`!  that's very helpful speed-up, why not add this `block height` improvement to Grin repo?  :-)

**Comment by DavidBurkett** [comment]
Mostly because we don't yet have precedent for changing underlying data storage format. So what might seem like a minor addition would in all likelihood lead to a huge discussion around backward compatibility. There are a few other improvements I'd like to make to Grin first that would lead to even larger performance gains, but there's a bit of an innovation freeze until after the hard fork :/

**Comment by garyyu** [comment]
I record this big improvement point in Grin repo   https://github.com/mimblewimble/grin/issues/2889.

> There are a few other improvements I'd like to make to Grin first that would lead to even larger performance gains

I'm very interested on the `other improvements`,  could you write down them one by one?

>  but there's a bit of an innovation freeze until after the hard fork 

Don't worry about that. These improvement are important but not urgent, you can submit them now but it's reasonable to wait for more discussions and reviews before going to one release:-)

**Comment by DavidBurkett** [comment]
> I'm very interested on the other improvements, could you write down them one by one?

I discussed some of them in the grin-dev channel with Jasper and others before the dev meeting Tuesday. The bulk of them boiled down to 3 things:

1) More specialized APIs. The v1 APIs were pretty generic, returned too much info, and queried very inefficiently  - look at how many times the same header is deserialized during wallet restore, for example. v2 may have improved this a lot actually, but I haven't checked.
2) Keep recent headers in memory. The common theme with our performance issues is deserialization, and in particular, headers. Grin++ is literally 10x faster at processing headers and 20x faster at wallet restores (I have done many benchmarks on both) because it doesn't have to keep deserializing old headers to load what it needs.
3) Only calculate master nonce on restore once per restore action, instead of once per output. I found out Tuesday that @jaspervdm actually implemented this in one of his PRs, so this one is already taken care of.
